packages/network-controller/src/NetworkController.ts

```diff
@@ -1063,205 +1475,441 @@ export class NetworkController extends BaseController<
-      existingAutoManagedNetworkClient.destroy();
+
+    const newNetworkConfiguration =
+      this.#determineNetworkConfigurationToPersist({
+        networkFields: fields,
+        networkClientOperations,
+      });
+    this.#registerNetworkClientsAsNeeded({
+      networkFields: fields,
+      networkClientOperations,
+      autoManagedNetworkClientRegistry,
+    });
+    this.update((state) => {
+      this.#updateNetworkConfigurations({
+        state,
+        mode: 'add',
+        networkFields: fields,
+        networkConfigurationToPersist: newNetworkConfiguration,
+      });
+    });
+
+    this.messagingSystem.publish(
+      `${controllerName}:networkAdded`,
+      newNetworkConfiguration,
+    );
+
+    return newNetworkConfiguration;
+  }
+
+  /**
+   * Updates the configuration for a previously stored network filed under the
+   * given chain ID, creating + registering new network clients to represent RPC
+   * endpoints that have been added and destroying + unregistering existing
+   * network clients for RPC endpoints that have been removed.
+   *
+   * Note that if `chainId` is changed, then all network clients associated with
+   * that chain will be removed and re-added, even if none of the RPC endpoints
+   * have changed.
+   *
+   * @param chainId - The chain ID associated with an existing network.
+   * @param fields - The object that describes the updates to the network/chain,
+   * including the new set of RPC endpoints which should front that chain.
+   * @param options - Options to provide.
+   * @param options.replacementSelectedRpcEndpointIndex - Usually you cannot
+   * remove an RPC endpoint that is being represented by the currently selected
+   * network client. This option allows you to specify another RPC endpoint
+   * (either an existing one or a new one) that should be used to select a new
+   * network instead.
+   * @returns The updated network configuration.
+   * @throws if `chainId` does not refer to an existing network configuration,
+   * if any part of `fields` would produce invalid state, etc.
+   * @see {@link NetworkConfiguration}
+   */
+  async updateNetwork(
+    chainId: Hex,
+    fields: UpdateNetworkFields,
+    {
+      replacementSelectedRpcEndpointIndex,
+    }: { replacementSelectedRpcEndpointIndex?: number } = {},
+  ): Promise<NetworkConfiguration> {
+    const existingNetworkConfiguration =
+      this.state.networkConfigurationsByChainId[chainId];
+
+    if (existingNetworkConfiguration === undefined) {
+      throw new Error(
+        `Could not update network: Cannot find network configuration for chain ${inspect(
+          chainId,
+        )}`,
+      );
+    }
+
+    const existingChainId = chainId;
+    const { chainId: newChainId, rpcEndpoints: setOfNewRpcEndpointFields } =
+      fields;
+
+    const autoManagedNetworkClientRegistry =
+      this.#ensureAutoManagedNetworkClientRegistryPopulated();
+
+    this.#validateNetworkFields({
```

Comment:
I noticed that we call this method in two places only, but we need to pass `autoManagedNetworkClientRegistry` and both `mode` and `errorMessagePrefix`.

I feel like this gives us the ability to extend its validation capability, but at the expense of a higher complexity for the caller: do you think that we would be able to derive `autoManagedNetworkClientRegistry` and `errorMessagePrefix` based on the `mode` alone?

Reply:
It's certainly possible to derive `errorMessagePrefix` inside of `#validateNetworkFields` instead of asking the caller to pass it. I've pushed commit [f25fb8c](https://github.com/MetaMask/core/pull/4286/commits/f25fb8c5693173b8147dba83aeb8152697af3619) to address this.

For `autoManagedNetworkClientRegistry`, we can't derive that, as we need it in `updateNetwork` ahead of time since we also need to pass it to `#registerClientsAsNeeded` and `#unRegisterClientsAsNeeded`.

Reply:
I see, that makes sense! Thanks for addressing `errorMessagePrefix`